### ISSUE TITLE
Mediator pattern and Pub/Sub clarifications

### ISSUE THREAD

**Issue by andrzejpolis** [addyosmani/essential-js-design-patterns#41]
I think a better job could be done of explaining the differences between these patterns in the book. I think I've seen you more clearly define them elsewhere but unsure if these have been brought back to EJDP.

This is your implementation of pub/sub

``` javascript
var pubsub = {};

(function(q) {

    var topics = {},
        subUid = -1;

    // Publish or broadcast events of interest
    // with a specific topic name and arguments
    // such as the data to pass along
    q.publish = function( topic, args ) {

        if ( !topics[topic] ) {
            return false;
        }

        var subscribers = topics[topic],
            len = subscribers ? subscribers.length : 0;

        while (len--) {
            subscribers[len].func( topic, args );
        }

        return this;
    };

    // Subscribe to events of interest
    // with a specific topic name and a
    // callback function, to be executed
    // when the topic/event is observed
    q.subscribe = function( topic, func ) {

        if (!topics[topic]) {
            topics[topic] = [];
        }

        var token = ( ++subUid ).toString();
        topics[topic].push({
            token: token,
            func: func
        });
        return token;
    };

    // Unsubscribe from a specific
    // topic, based on a tokenized reference
    // to the subscription
    q.unsubscribe = function( token ) {
        for ( var m in topics ) {
            if ( topics[m] ) {
                for ( var i = 0, j = topics[m].length; i < j; i++ ) {
                    if ( topics[m][i].token === token) {
                        topics[m].splice( i, 1 );
                        return token;
                    }
                }
            }
        }
        return this;
    };
}( pubsub ));
```

Here goes mediator:

``` javascript
var mediator = (function(){

    // Storage for topics that can be broadcast or listened to
    var topics = {};

    // Subscribe to a topic, supply a callback to be executed
    // when that topic is broadcast to
    var subscribe = function( topic, fn ){

        if ( !topics[topic] ){ 
          topics[topic] = [];
        }

        topics[topic].push( { context: this, callback: fn } );

        return this;
    };

    // Publish/broadcast an event to the rest of the application
    var publish = function( topic ){

        var args;

        if ( !topics[topic] ){
          return false;
        } 

        args = Array.prototype.slice.call( arguments, 1 );
        for ( var i = 0, l = topics[topic].length; i < l; i++ ) {

            var subscription = topics[topic][i];
            subscription.callback.apply( subscription.context, args );
        }
        return this;
    };

    return {
        Publish: publish,
        Subscribe: subscribe,
        installTo: function( obj ){
            obj.subscribe = subscribe;
            obj.publish = publish;
        }
    };

}());
```

They are identical. In fact they both look exactly like correct pub/sub pattern. 
Mediator pattern if I understand correctly works as follows:
- mediator knows all components - components may inform mediator about it's existence during creation - this is preferred way.
- components informs mediator
- mediator decides what to do -  it decides which components need updates and in what form.

In other words - pub/sub is not adequate here. We don't wont components to subscribe. They don't decide in this strategy. They should only inform mediator about information from their scope. They also should expose some api - mediator uses this api to control components.

**Comment by nickolaslanza** [comment]
Hi Addy,

I'm also enjoying reading EJSDP and am really enjoying learning different design patterns, but am having a hard time reading through the current version of the Mediator/Event Aggregator Pattern in this [book](https://addyosmani.com/resources/essentialjsdesignpatterns/book/#mediatorpatternjavascript). Your [link](https://addyosmani.com/backbone-fundamentals/#event-aggregators-and-mediators) to the new reference does help. What helped me further simply the idea of Event Aggregator vs Mediator was to read [this](https://coderwall.com/p/n_nvmw/difference-between-event-aggregator-and-mediator).
